File: react-client/src/suggestion-tree.ts
type TreeNode = {
  _isLeaf: boolean;
  childrens: Record<string, TreeNode>;
};

export default class SuggestionTree {
  private suggestions: string[] = [];
  private root: TreeNode | null = null;

  static emptyNode(): TreeNode {
    return { _isLeaf: false, childrens: {} };
  }

  append(word: string) {
    if (!this.root) this.root = SuggestionTree.emptyNode();
    let ptr = this.root;

    for (const letter of word) {
      if (!(letter in ptr.childrens)) {
        ptr.childrens[letter] = SuggestionTree.emptyNode();
      }
      ptr = ptr.childrens[letter];
    }

    ptr._isLeaf = true;
  }

  private find(word: string): TreeNode | null {
    let ptr = this.root;

    for (const letter of word) {
      if (letter in ptr!.childrens) {
        ptr = ptr!.childrens[letter];
      } else {
        return null;
      }
    }

    return ptr;
  }

  private fill(node: TreeNode, word: string) {
    if (node._isLeaf) {
      this.suggestions.push(word);
      return;
    }

    for (const letter in node.childrens) {
      this.fill(node.childrens[letter], word + letter);
    }
  }

  suggest(word: string) {
    const node = this.find(word);
    node && this.fill(node, word);

    return this.suggestions;
  }

  flush() {
    this.suggestions = [];
  }
}

File: wasm/src/command/suggestion_tree.rs
use std::collections::HashMap;

const COMMANDS: [&str; 3] = ["Filter", "Average", "Sum"];

#[derive(Default, Debug)]
pub struct TreeNode {
    is_leaf: bool,
    childrens: HashMap<u8, Box<TreeNode>>,
}

#[derive(Default, Debug)]
pub struct SuggestionTree {
    root: Option<TreeNode>,
}

impl SuggestionTree {
    pub fn append(&mut self, word: &str) {
        let mut ptr = self.root.get_or_insert_default();

        for byte in word.as_bytes() {
            ptr = ptr
                .childrens
                .entry(*byte)
                .or_insert_with(|| Box::new(TreeNode::default()));
        }

        ptr.is_leaf = true;
    }

    fn find(&self, word: &str) -> Option<&TreeNode> {
        let mut ptr = self.root.as_ref().unwrap();

        for byte in word.as_bytes() {
            ptr = ptr.childrens.get(byte)?;
        }

        Some(ptr)
    }

    fn get_suggestions(node: &TreeNode, word: &str, ret: &mut Vec<String>) {
        let mut s = word.to_string();

        if node.is_leaf {
            ret.push(s);
        } else {
            for (byte, node) in node.childrens.iter() {
                s.push(*byte as char);
                Self::get_suggestions(node, &s, ret);
            }
        }
    }

    pub fn suggest(&self, word: &str) -> Vec<String> {
        let mut ret = Vec::new();
        let node = self.find(word).unwrap();
        Self::get_suggestions(node, word, &mut ret);

        ret
    }

    pub fn base_tree() -> Self {
        let mut tree = Self::default();
        COMMANDS.iter().for_each(|&c| tree.append(c));
        tree
    }
}

#[cfg(test)]
mod test {
    use super::SuggestionTree;

    #[test]
    fn test() {
        let mut tree = SuggestionTree::default();
        tree.append("foo");
        tree.append("bar");
        tree.append("baz");

        let suggestions = tree.suggest("b");
        assert_eq!(suggestions.len(), 2);
    }
}
